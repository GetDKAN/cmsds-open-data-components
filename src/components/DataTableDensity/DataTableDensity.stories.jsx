import React, { useState } from 'react';
import DataTableDensity from './index';

const meta = {
  title: 'Components/DataTableDensity',
  component: DataTableDensity,
  parameters: {
    layout: 'padded',
    docs: {
      description: {
        component: `\nThe DataTableDensity component provides a dropdown to select display density for a data table.\n        `,
      },
    },
  },
  argTypes: {
    setTablePadding: { control: false, description: 'Callback to set table padding.' },
    tablePadding: { control: 'text', description: 'Current table padding value.' },
  },
  tags: ['autodocs'],
};

export default meta;

export const Default = {
  render: (args) => {
    const [tablePadding, setTablePadding] = useState('ds-u-padding-y--1');
    return <DataTableDensity setTablePadding={setTablePadding} tablePadding={tablePadding} />;
  },
  args: {
    tablePadding: 'ds-u-padding-y--1',
  },
};
