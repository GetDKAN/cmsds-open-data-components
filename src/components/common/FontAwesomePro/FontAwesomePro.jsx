import React, { useRef, useEffect } from 'react'
import PropTypes from 'prop-types'
import svgIndex from '../../../assets/icons'

const FontAwesomePro = (props) => {
  const {
    'aria-hidden': ariaHidden,
    fill = '#323A45',
    height = 16,
    icon = 'long-arrow-right',
    size,
    width = 16
  } = props

  // Set fill on path
  const ref = useRef()
  useEffect(() => {
    if (fill && ref.current && ref.current.firstElementChild) {
      ref.current.firstElementChild.setAttribute('fill', fill)
    }
    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [icon])

  // Set default svg attribtues
  // const iconAttr = { /////////////////
  //   'aria-hidden': ariaHidden,
  //   className: `fa fa-${icon}`,
  //   focusable: 'false',
  //   height: size || height,
  //   role: 'img',
  //   width: size || width
  // }

  // Get svgIndex export array values
  const svgList = svgIndex.values() ////////

  // Loop through svgIndex export array
  // const iconMap = (ico) => { //////////////
  //   for (const svg of svgList) {
  //     const key = Object.keys(svg)[0]
  //     if (key === ico) {
  //       // Get value from svgIndex, which is a ReactComponent
  //       const Value = Object.values(svg)[0]
  //       return <Value {...iconAttr} ref={ref} aria-label={icon.split('-').join(' ')} />
  //     }
  //   }
  // }

  // return ( ///////
  //   iconMap(icon)
  // )
}

FontAwesomePro.propTypes = {
  /**
   * Hide icon from screen reader
   */
  'aria-hidden': PropTypes.bool,
  /**
   * Icon color code string (hex, rgb, string, etc)
   */
  fill: PropTypes.string,
  /**
   * SVG outer pixel height - Overwritten by 'size' prop if provided
   */
  height: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * FontAwesome icon name
   */
  icon: PropTypes.string,
  /**
   * SVG pixel size - Applies to height and width equally
   */
  size: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
  /**
   * SVG outer pixel width - Overwritten by 'size' prop if provided
   */
  width: PropTypes.oneOfType([PropTypes.string, PropTypes.number])
}
FontAwesomePro.displayName = 'FontAwesomePro'
export default FontAwesomePro
