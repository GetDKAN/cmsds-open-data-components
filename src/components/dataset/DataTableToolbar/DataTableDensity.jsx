import React from 'react'
import PropTypes from 'prop-types'
import DataIcon from './DataIcon'

const DataTableDensity = ({
  active,
  items = [
    { icon: <DataIcon name='density-1' height={20} width={20} icon='density-1' fill='#666666' />, text: 'expanded', value: 'density-1' },
    { icon: <DataIcon name='density-2' height={20} width={20} icon='density-2' fill='#666666' />, text: 'normal', value: 'density-2' },
    { icon: <DataIcon name='density-3' height={20} width={20} icon='density-3' fill='#666666' />, text: 'tight', value: 'density-3' }
  ],
  densityChange,
  className = 'data-table-density',
  screenReaderClass = 'sr-only sr-only-focusable',
  title = 'Display Density'
}) => (
  <div className={className}>
    <span className='density-buttons-title'>{title}</span>
    <div className='density-buttons'>
      {items.map((item) => {
        let srClass = screenReaderClass
        if (!item.icon) {
          srClass = ''
        }
        return (
          <button
            type='button'
            key={item.text}
            onMouseUp={() => densityChange(item.value)}
            onTouchEnd={() => densityChange(item.value)}
            onClick={() => densityChange(item.value)}
            title={item.text}
            className={active === item.text ? 'active' : ''}
            aria-label={`${title}, ${item.text}${active === item.text ? ' - selected' : ''}`}
          >
            {item.icon &&
              (
                <>
                  {item.icon}
                </>
              )}
            <span className={srClass}>{item.text}</span>
          </button>
        )
      })}
    </div>
  </div>
)

DataTableDensity.propTypes = {
  /**
   * Button active state class.
   * Text of active item. Ex. 'Normal'
   */
  active: PropTypes.string,
  /**
   * Button click callback function
   */
  densityChange: PropTypes.func.isRequired,
  /**
   * Custom class(es) used for accessibility purposes.
   * Applied to button text span
   */
  screenReaderClass: PropTypes.string,
  /**
   * Custom class(es) applied to wrapping element
   */
  className: PropTypes.string,
  /**
   * Array of button data
   */
  items: PropTypes.arrayOf(PropTypes.shape({
    icon: PropTypes.node,
    text: PropTypes.string,
    value: PropTypes.string
  })),
  /**
   * Button group title text
   */
  title: PropTypes.string
}

DataTableDensity.displayName = 'DataTableDensity'
export default DataTableDensity
