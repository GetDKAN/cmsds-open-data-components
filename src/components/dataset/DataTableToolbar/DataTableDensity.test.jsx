import { render, fireEvent } from '@testing-library/react'
import renderer from 'react-test-renderer'
import DataTableDensity from './DataTableDensity'
import FontAwesomePro from '../../common/FontAwesomePro/FontAwesomePro'

const componentArgs = {
  title: 'Display Density',
  items: [
    {
      icon: (
        <FontAwesomePro icon='density-1' aria-hidden />
      ),
      text: 'Expanded',
      value: 'density-1'
    },
    {
      icon: (
        <FontAwesomePro icon='density-2' aria-hidden />
      ),
      text: 'Normal',
      value: 'density-2'
    },
    {
      icon: (
        <FontAwesomePro icon='density-3' aria-hidden />
      ),
      text: 'Compact',
      value: 'density-3'
    }
  ],
  densityChange: vi.fn(),
  screenReaderClass: 'sr-only sr-only-focusable',
  className: 'data-table-density',
  active: 'Normal'
}

describe('DataTableDensity component.', () => {
  it('Matches snapshot.', () => {
    const renderedDataTableDensity = renderer.create(
      <DataTableDensity {...componentArgs} />
    ).toJSON()

    expect(renderedDataTableDensity).toMatchSnapshot()
  })

  it('\'densityChange\' prop is called when buttons are clicked.', () => {
    const densityChange = vi.fn()

    const { container } = render(
      <DataTableDensity
        {...componentArgs}
        densityChange={densityChange}
      />
    )

    componentArgs.items.forEach((button) => {
      fireEvent.click(container.querySelector(`button[aria-label*="${componentArgs.title}, ${button.text}"]`))
      expect(densityChange).toHaveBeenCalled()

      fireEvent.mouseUp(container.querySelector(`button[aria-label*="${componentArgs.title}, ${button.text}"]`))
      expect(densityChange).toHaveBeenCalled()

      fireEvent.touchEnd(container.querySelector(`button[aria-label*="${componentArgs.title}, ${button.text}"]`))
      expect(densityChange).toHaveBeenCalled()
    })
  })

  it('Screen reader class is not applied when buttons aren\'t provided with icons.', () => {
    const { container } = render(
      <DataTableDensity
        {...componentArgs}
        items={[
          {
            text: 'Expanded',
            value: 'density-1'
          },
          {
            text: 'Normal',
            value: 'density-2'
          },
          {
            text: 'Compact',
            value: 'density-3'
          }
        ]}
      />
    )

    componentArgs.items.forEach((button) => {
      expect(container.querySelector(`button[aria-label*="${componentArgs.title}, ${button.text}"] > span`)).not.toHaveClass(componentArgs.screenReaderClass)
    })
  })
})
