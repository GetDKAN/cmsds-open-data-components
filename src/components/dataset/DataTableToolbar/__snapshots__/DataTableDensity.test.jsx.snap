// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DataTableDensity component. > Matches snapshot. 1`] = `
<div
  className="data-table-density"
>
  <span
    className="density-buttons-title"
  >
    Display Density
  </span>
  <div
    className="density-buttons"
  >
    <button
      aria-label="Display Density, Expanded"
      className=""
      onClick={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      title="Expanded"
      type="button"
    >
      <svg
        aria-hidden={true}
        aria-label="density 1"
        className="fa fa-density-1"
        focusable="false"
        height={16}
        role="img"
        viewBox="0 0 512 512"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M496,384 C504.836556,384 512,391.163444 512,400 L512,496 C512,504.836556 504.836556,512 496,512 L16,512 C7.163444,512 1.082166e-15,504.836556 0,496 L0,400 C-1.082166e-15,391.163444 7.163444,384 16,384 L496,384 Z M496,192 C504.836556,192 512,199.163444 512,208 L512,304 C512,312.836556 504.836556,320 496,320 L16,320 C7.163444,320 1.082166e-15,312.836556 0,304 L0,208 C-1.082166e-15,199.163444 7.163444,192 16,192 L496,192 Z M496,0 C504.836556,-1.623249e-15 512,7.163444 512,16 L512,112 C512,120.836556 504.836556,128 496,128 L16,128 C7.163444,128 1.082166e-15,120.836556 0,112 L0,16 C-1.082166e-15,7.163444 7.163444,1.623249e-15 16,0 L496,0 Z"
        />
      </svg>
      <span
        className="sr-only sr-only-focusable"
      >
        Expanded
      </span>
    </button>
    <button
      aria-label="Display Density, Normal - selected"
      className="active"
      onClick={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      title="Normal"
      type="button"
    >
      <svg
        aria-hidden={true}
        aria-label="density 2"
        className="fa fa-density-2"
        focusable="false"
        height={16}
        role="img"
        viewBox="0 0 512 512"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M496,416 C504.836556,416 512,423.163444 512,432 L512,496 C512,504.836556 504.836556,512 496,512 L16,512 C7.163444,512 1.082166e-15,504.836556 0,496 L0,432 C-1.082166e-15,423.163444 7.163444,416 16,416 L496,416 Z M496,277 C504.836556,277 512,284.163444 512,293 L512,357 C512,365.836556 504.836556,373 496,373 L16,373 C7.163444,373 1.082166e-15,365.836556 0,357 L0,293 C-1.082166e-15,284.163444 7.163444,277 16,277 L496,277 Z M496,139 C504.836556,139 512,146.163444 512,155 L512,219 C512,227.836556 504.836556,235 496,235 L16,235 C7.163444,235 1.082166e-15,227.836556 0,219 L0,155 C-1.082166e-15,146.163444 7.163444,139 16,139 L496,139 Z M496,0 C504.836556,-1.623249e-15 512,7.163444 512,16 L512,80 C512,88.836556 504.836556,96 496,96 L16,96 C7.163444,96 1.082166e-15,88.836556 0,80 L0,16 C-1.082166e-15,7.163444 7.163444,1.623249e-15 16,0 L496,0 Z"
        />
      </svg>
      <span
        className="sr-only sr-only-focusable"
      >
        Normal
      </span>
    </button>
    <button
      aria-label="Display Density, Compact"
      className=""
      onClick={[Function]}
      onMouseUp={[Function]}
      onTouchEnd={[Function]}
      title="Compact"
      type="button"
    >
      <svg
        aria-hidden={true}
        aria-label="density 3"
        className="fa fa-density-3"
        focusable="false"
        height={16}
        role="img"
        viewBox="0 0 512 512"
        width={16}
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M496,448 C504.836556,448 512,455.163444 512,464 L512,496 C512,504.836556 504.836556,512 496,512 L16,512 C7.163444,512 1.082166e-15,504.836556 0,496 L0,464 C-1.082166e-15,455.163444 7.163444,448 16,448 L496,448 Z M496,336 C504.836556,336 512,343.163444 512,352 L512,384 C512,392.836556 504.836556,400 496,400 L16,400 C7.163444,400 1.082166e-15,392.836556 0,384 L0,352 C-1.082166e-15,343.163444 7.163444,336 16,336 L496,336 Z M496,224 C504.836556,224 512,231.163444 512,240 L512,272 C512,280.836556 504.836556,288 496,288 L16,288 C7.163444,288 1.082166e-15,280.836556 0,272 L0,240 C-1.082166e-15,231.163444 7.163444,224 16,224 L496,224 Z M496,112 C504.836556,112 512,119.163444 512,128 L512,160 C512,168.836556 504.836556,176 496,176 L16,176 C7.163444,176 1.082166e-15,168.836556 0,160 L0,128 C-1.082166e-15,119.163444 7.163444,112 16,112 L496,112 Z M496,0 C504.836556,-1.623249e-15 512,7.163444 512,16 L512,48 C512,56.836556 504.836556,64 496,64 L16,64 C7.163444,64 1.082166e-15,56.836556 0,48 L0,16 C-1.082166e-15,7.163444 7.163444,1.623249e-15 16,0 L496,0 Z"
        />
      </svg>
      <span
        className="sr-only sr-only-focusable"
      >
        Compact
      </span>
    </button>
  </div>
</div>
`;
